src/util/hashFetch/expoRetry.js

```diff
@@ -0,0 +1,118 @@
+const util = require('util');
+const path = require('path');
+
+import { readJson as fsReadJson, writeJson as fsWriteJson } from 'fs-extra';
+const fsExists = util.promisify(fs.stat);
+
+import { getHashFetchPath } from '../host';
+
+// Handle retries with exponential delay for failed download attempts
+export default class ExpoRetry {
+  static instance = null;
+
+  // We store the URL to allow GH Hint URL updates for a given hash
+  // to take effect immediately, and to disregard past failed attempts at
+  // getting this file from other URLs.
+  // {
+  //   [`${hash}:${url}`]: {
+  //     attempts: [
+  //       {timestamp: 1529598899017},
+  //       ...
+  //     ]
+  //   },
+  //   ...
+  // }
+  failHistory = {};
+
+  // If true, the failHistory var was updated and needs to be written to disk
+  needWrite = false;
+  writeQueue = Promise.resolve();
+
+  static get () {
+    if (!ExpoRetry.instance) {
+      ExpoRetry.instance = new ExpoRetry();
+    }
+
+    return ExpoRetry.instance;
+  }
+
+  _getFilePath () {
+    return path.join(getHashFetchPath(), 'fail_history.json');
+  }
+
+  _getKey (hash, url) {
+    return `${hash}:${url}`;
+  }
+
+  load () {
+    const filePath = this._getFilePath();
+
+    return fsExists(filePath)
+        .then(() =>
+          fsReadJson(filePath)
+            .catch(e => {
+              throw new Error(`Couldn't parse JSON for ExpoRetry file ${filePath} ${e}`);
+            })
+        )
+        .then(failHistory => {
+          this.failHistory = failHistory;
+        })
+        .catch(() => fsWriteJson(filePath, this.failHistory));
+  }
+
+  canAttemptDownload (hash, url) {
+    const key = this._getKey(hash, url);
+
+    // Never tried downloading the file
+    if (!(key in this.failHistory) || !this.failHistory[key].attempts.length) {
+      return true;
+    }
+
+    // Already failed at downloading the file: check if we can retry now
+    // Delay starts at 30 seconds, max delay is 23 days
+    const retriesCount = this.failHistory[key].attempts.length - 1;
+    const latestAttemptTimestamp = this.failHistory[key].attempts.slice(-1)[0].timestamp;
+    const earliestNextAttemptTimestamp = latestAttemptTimestamp + Math.pow(2, Math.min(16, retriesCount)) * 30000;
+
+    if (Date.now() > earliestNextAttemptTimestamp) {
+      return true;
+    }
+
+    return false;
+  }
+
+  registerFailedAttempt (hash, url) {
+    const key = this._getKey(hash, url);
+
+    this.failHistory[key] = this.failHistory[key] || { attempts: [] };
+    this.failHistory[key].attempts.push({ timestamp: Date.now() });
+
+    // Once the ongoing write is finished, write anew with the updated contents
+    this.needWrite = true;
+    this.writeQueue = this.writeQueue.then(() => {
+      if (this.needWrite) {
+        // Skip subsequent promises, considering we are writing the latest value
+        this.needWrite = false;
+
+        return fsWriteJson(this._getFilePath(), this.failHistory);
+      }
+    });
```

Comment:
Since we are not handling errors here it seems that if it fails once it will never be attempted again, right? Should we at least print a warning if that happens?

Reply:
Nice catch! (or rather, missing catch :P) I don't think we need to never attempt writing again if it failed once. I'll add a catch